java-checks/src/main/java/org/sonar/java/symexec/ExecutionState.java

```diff
@@ -220,4 +223,29 @@ private void mergeBooleanConstraints(Iterable<ExecutionState> states) {
     return result;
   }
 
+  void invalidateRelationsOnSymbol(VariableSymbol symbol) {
+    Multimap<Symbol.VariableSymbol, Symbol.VariableSymbol> pairs = HashMultimap.create();
+    for (ExecutionState current = this; current != null; current = current.parentState) {
+      findRelationPairs(current, pairs);
+    }
+    for (Map.Entry<Symbol.VariableSymbol, Symbol.VariableSymbol> cell : pairs.entries()) {
+      if (getRelation(cell.getKey(), cell.getValue()) != SymbolicRelation.UNKNOWN) {
+        setRelation(cell.getKey(), SymbolicRelation.UNKNOWN, cell.getValue());
+      }
+    }
+  }
+
+  private static void findRelationPairs(ExecutionState state, Multimap<Symbol.VariableSymbol, Symbol.VariableSymbol> pairs) {
```

Comment:
IMO this should return a Multimap that should be put in result using putAll method. Hence it will not modify input parameter (which will be only an execution state) and you will have a pure function.